Sources/NIOPosix/Bootstrap.swift

```diff
@@ -521,6 +521,199 @@ public final class ServerBootstrap {
+    ///                )
+    ///            }
+    ///        }
+    ///    ) { channel in
+    ///        print("Handling new connection")
+    ///        await self.handleConnection(channel: channel)
+    ///        print("Done handling connection")
+    ///    }
+    ///
+    /// ```
+    ///
+    /// - Parameters:
+    ///   - target: The ``BindTarget`` to use.
+    ///   - serverBackPressureStrategy: The back pressure strategy used by the server socket channel.
+    ///   - childChannelInitializer: A closure to initialize the channel. The return value of this closure is used in the `onConnection`
+    ///                              closure.
+    ///   - onConnection: A closure to handle the connection. Use the channel's `inbound` property to read from
+    ///                   the connection and channel's `outbound` to write to the connection.
+    ///
+    /// - Note: The bind method respects task cancellation which will force close the server. If you want to gracefully
+    ///         shut-down use the quiescing helper approach as outlined above.
+    @available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *)
+    public func bind<Inbound: Sendable, Outbound: Sendable>(
+        target: BindTarget,
+        serverBackPressureStrategy: NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark? = nil,
+        childChannelInitializer: @escaping @Sendable (Channel) -> EventLoopFuture<NIOAsyncChannel<Inbound, Outbound>>,
+        _ onConnection: @escaping @Sendable (
+            _ channel: NIOAsyncChannel<Inbound, Outbound>
+        ) async -> ()
+    ) async throws {
+        let channel = try await self.makeConnectedChannel(
+            target: target,
+            serverBackPressureStrategy: serverBackPressureStrategy,
+            childChannelInitializer: childChannelInitializer
+        )
+
+        try await withTaskCancellationHandler {
+            try await channel.executeThenClose { inbound, outbound in
+                // we need to dance the result dance here, since we can't throw from the
+                // withDiscardingTaskGroup closure.
+                let result = await withDiscardingTaskGroup { group -> Result<Void, any Error> in
+                    do {
+                        try await channel.executeThenClose { inbound in
+                            for try await connectionChannel in inbound {
+                                group.addTask {
+                                    do {
+                                        try await connectionChannel.executeThenClose { _, _ in
+                                            await onConnection(connectionChannel)
+                                        }
+                                    } catch {
+                                        // ignore single connection failures
+                                    }
+                                }
+                            }
+                        }
+                        return .success(())
```

Comment:
Does this implementation need to know about quiescing and/or graceful shutdown?

What's the expected behavior when a server stops accepting clients I.E. ServerBootstrap closes? Do all child tasks cancel, or do we wait for them to shut down? This kinda plays into graceful shutdown from Service Lifecycle. Can a connection continue to live while the parent channel is closed?